dozer-sink-aerospike/src/lib.rs

```diff
@@ -0,0 +1,960 @@
+            );
+        }
+    }
+
+    unsafe fn init_key(
+        &self,
+        key: *mut as_key,
+        record: &Record,
+        allocated_strings: &mut Vec<String>,
+    ) -> Result<(), AerospikeSinkError> {
+        let key_field = &record.values[self.primary_index];
+        unsafe {
+            match key_field {
+                Field::UInt(v) => {
+                    as_key_init_int64(
+                        key,
+                        self.namespace.as_ptr(),
+                        self.set_name.as_ptr(),
+                        *v as i64,
+                    );
+                }
+                Field::Int(v) => {
+                    as_key_init_int64(key, self.namespace.as_ptr(), self.set_name.as_ptr(), *v);
+                }
+                Field::U128(v) => self.set_str_key(key, v.to_string(), allocated_strings),
+                Field::I128(v) => self.set_str_key(key, v.to_string(), allocated_strings),
+                Field::Decimal(v) => self.set_str_key(key, v.to_string(), allocated_strings),
+                // For keys, we need to allocate a new CString, because there is no
+                // API to set a key to a string that's not null-terminated. For bin
+                // values, we can. XXX: possible point for optimization
+                Field::Text(string) | Field::String(string) => {
+                    // Casting to mut is safe. The pointer only needs to be mut so it
+                    // can be deallocated if the `free` parameter is true
+                    let bytes =
+                        as_bytes_new_wrap(string.as_ptr() as *mut u8, string.len() as u32, false);
+                    (*bytes).type_ = as_bytes_type_e_AS_BYTES_STRING;
+                    as_key_init_value(
+                        key,
+                        self.namespace.as_ptr(),
+                        self.set_name.as_ptr(),
+                        bytes as *const _ as *const as_key_value,
+                    );
+                }
+                Field::Binary(v) => {
+                    as_key_init_rawp(
+                        key,
+                        self.namespace.as_ptr(),
+                        self.set_name.as_ptr(),
+                        v.as_ptr(),
+                        v.len() as u32,
+                        false,
+                    );
+                }
+                //
+                Field::Timestamp(v) => self.set_str_key(
+                    key,
+                    // Use a delayed formatting to RFC3339 so we don't have to allocate an
+                    // intermediate rust String
+                    v.to_rfc3339(),
+                    allocated_strings,
+                ),
+                // Date's display implementation is RFC3339 compatible
+                Field::Date(v) => self.set_str_key(key, v.to_string(), allocated_strings),
+                // FIXME: Not sure how time unit should be used, but I'm pretty sure ignoring it
```

Comment:
The unit is a hint, so when we convert to external systems that have limited duration precision, we know if it's possible or not.

For example, if the external system only supports duration in milliseconds, and our unit is nanoseconds, we know we have to either report an error or lose precision.

Reply:
Alright, so in this case it's fine to ignore it, as we're pushing a full-resolution string anyway